### PR TITLE
Merge sparse adapter

### PR DESCRIPTION
**Pull request by fverac**
PR to merge Samin's work on a SparseMaskAdapter into the library. This should supercede the existing sparse mask implementations

### CODE REVIEW

File: mttl/arguments.py [hunk 1]
```diff
@@ -344,6 +344,11 @@ class TrainingArgs(DataArgs):
     exp_name: str = None
     expert_name: str = None
 
+    K: int = (
```
Comment (fverac) [inline mttl/arguments.py:347]: I think these are removable, will try in next commit

**Comment by fverac** [comment]
I think we possibly exclude any changes/files in projects dir

File: mttl/models/modifiers/sparse_mask_old.py
Comment (SaminYeasar) [inline mttl/models/modifiers/sparse_mask_old.py:1]: we are not using this script

File: mttl/models/modifiers/sparse_mask.py [hunk 2]
```diff
@@ -1,121 +1,779 @@
...
+    module = module.to('cpu')
+    # if save_mask_indx: 
+    #     import h5py
+    #     import os
+    #     # Ensure the directory exists or create it if not
+    #     os.makedirs('saved_masks', exist_ok=True)
+    #     f_name = f'saved_masks/{dm.config.finetune_task_name}'
+    #     np.savez_compressed(f'{f_name}.npz', arr=mask_indx)
+    #     # with h5py.File(f'{f_name}.h5', 'w') as f:
+    #     #     f.create_dataset('data', data=mask_indx, compression='gzip')
```
Comment (SaminYeasar) [inline mttl/models/modifiers/sparse_mask.py]: please remove the commented-out part of the code

File: mttl/models/modifiers/sparse_mask.py [hunk 3]
```diff
@@ -1,121 +1,779 @@
...
+#     def forward(self, input):
+#         # `weight_mask` requires to sync with `weight` device, as default only looks module 
+        
+#         self.confirm_sync_device()  # TODO, need to remove this, it should only require once before training
+
+#         weight = self.layer.weight.flatten().scatter_add(0, self.sparse_layer.weight_mask, self.sparse_layer.weight)
+#         weight = weight.to(self.layer.weight.device)
+#         weight = weight.reshape_as(self.layer.weight)
+
+#         return torch.nn.functional.linear(input, weight, self.layer.bias + self.sparse_layer.bias)
```
Comment (SaminYeasar) [inline mttl/models/modifiers/sparse_mask.py]: please remove this commented-out function, we are not using it.

File: mttl/models/hf/callbacks.py [hunk 4]
```diff
@@ -11,6 +11,53 @@
...
+                # Update mask
+                self.update_mask(pl_module, batch)
+                self.update_counter = 0  # Reset counter for next interval
+                # save mask: if needed, if need to upload to hf, use `on_train_epoch_end` or `on_train_end` to reduce number of request to hf server
+                # f_name = f'{self.save_mask_dir}/{self.task_name}_mask'
+                # # f_name = f'{self.save_mask_dir}/mask'
+                # save_mask(pl_module, f_name)
+                
+                # TODO: don't use this during training, use only required during debugging
+                #self.on_train_end(trainer, pl_module)
```
Comment (SaminYeasar) [inline mttl/models/hf/callbacks.py]: Please remove this commented-out section

File: mttl/models/hf/callbacks.py [hunk 5]
```diff
@@ -11,6 +11,53 @@
...
+        save_mask(pl_module, f_name)
+        
+        #from mttl.models.modifiers.sparse_mask import compress_sparse_2D_weight
+        #compress_sparse_2D_weight(pl_module)
+
+
+    # def on_train_epoch_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+    #     f_name = f'{self.save_mask_dir}/{self.task_name}_mask'
+    #     # f_name = f'{self.save_mask_dir}/mask'
+    #     save_mask(pl_module, f_name)
```
Comment (SaminYeasar) [inline mttl/models/hf/callbacks.py]: Please remove this commented-out section

File: projects/sparse_finetuning/eval_library.py [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+                for threshold in [1e-10, 1e-5, 1e-3, 1e-2, 1e-1]:
+                    # Effective rank: Count singular values greater than threshold * largest_singular_value
+                    largest_singular_value = s_b[0]
+                    effective_rank_base = np.sum(s_b > threshold * largest_singular_value)
+
+                    #largest_singular_value = s_e[0]
+                    effective_rank_expert = np.sum(s_e > threshold * largest_singular_value)
+                    print(f'{w_name} {threshold} base | expert :', effective_rank_base, effective_rank_expert)
+
+    print('done')
```
Comment (SaminYeasar) [inline projects/sparse_finetuning/eval_library.py:564]: Please remove this commented-out function, we are not using this

File: projects/sparse_finetuning/eval_library.py [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+
+if __name__ == "__main__":
+    args = ExpertConfig.parse()
+    run_eval(args)
+    #compare_svd(args)
```
Comment (SaminYeasar) [inline projects/sparse_finetuning/eval_library.py:569]: please remove the commented-out function call

File: projects/sparse_finetuning/eval_library.py [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+    #selection = ['duorc_ParaphraseRC_extract_answer']
+    # selection = ['duorc_ParaphraseRC_extract_answer',
+    #               'wiki_qa_Topic_Prediction_Question_and_Answer_Pair',
+    #               'duorc_ParaphraseRC_build_story_around_qa',
+    #               'cos_e_v1_11_description_question_option_id','adversarial_qa_droberta_answer_the_following_q']
+    # selection = ['wiqa_what_is_the_final_step_of_the_following_process', 'sciq_Multiple_Choice', 'adversarial_qa_droberta_answer_the_following_q', 'duorc_SelfRC_question_answering', 'cos_e_v1_11_description_question_option_id', 'wiki_qa_Is_This_True_', 'quail_description_context_question_text', 'wiki_hop_original_explain_relation', 'duorc_ParaphraseRC_build_story_around_qa', 'yelp_polarity_reviews_0_2_0', 'squad_v1_1_3_0_0', 'web_questions_potential_correct_answer', 'quoref_Found_Context_Online', 'quoref_Given_Context_Answer_Question', 'web_questions_get_the_answer', 'cot_sensemaking', 'wiki_qa_Topic_Prediction_Question_and_Answer_Pair', 'dbpedia_14_given_list_what_category_does_the_paragraph_belong_to', 'duorc_ParaphraseRC_extract_answer', 'super_glue_rte_1_0_2']
```
Comment (SaminYeasar) [inline projects/sparse_finetuning/eval_library.py:289]: Please remove this commented-out part of the code

File: projects/sparse_finetuning/eval_library.py [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+
+    #args.library_id = 'local://library/phi_3_finetune_lora/'
+    #args.merge_or_route = "uniform"   # for sparse weight merge
+
+    # args.library_id = 'local://library/phi_3_finetune_0winit_noisy_sparse_kr_0.002/'
+    # args.library_id = 'local://library/phi_3_finetune_0.002_nr_0.02/'
+    # args.merge_or_route = "uniform_sparse_weight"   # for sparse weight merge
+    
+ 
+    # args.pipeline_eval_tasks = "in_distribution"
```
Comment (SaminYeasar) [inline projects/sparse_finetuning/eval_library.py:279]: please remove the commented part of the code

File: mttl/models/lightning/callbacks.py [hunk 10]
```diff
@@ -21,9 +21,56 @@
...
+        save_mask(pl_module, f_name)
+        
+        #from mttl.models.modifiers.sparse_mask import compress_sparse_2D_weight
+        #compress_sparse_2D_weight(pl_module)
+
+
+    # def on_train_epoch_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+    #     f_name = f'{self.save_mask_dir}/{self.task_name}_mask'
+    #     # f_name = f'{self.save_mask_dir}/mask'
+    #     save_mask(pl_module, f_name)
```
Comment (SaminYeasar) [inline mttl/models/lightning/callbacks.py]: please remove the commented-out part of the code

File: mttl/models/lightning/callbacks.py [hunk 11]
```diff
@@ -21,9 +21,56 @@
...
+                # Update mask
+                self.update_mask(pl_module, batch)
+                self.update_counter = 0  # Reset counter for next interval
+                # save mask: if needed, if need to upload to hf, use `on_train_epoch_end` or `on_train_end` to reduce number of request to hf server
+                # f_name = f'{self.save_mask_dir}/{self.task_name}_mask'
+                # # f_name = f'{self.save_mask_dir}/mask'
+                # save_mask(pl_module, f_name)
+                
+                # TODO: don't use this during training, use only required during debugging
+                #self.on_train_end(trainer, pl_module)
```
Comment (SaminYeasar) [inline mttl/models/lightning/callbacks.py]: please remove the commented-out part of the code

**Review by SaminYeasar — CHANGES_REQUESTED** [review]
Requested to remove commented-out code blocks and functions that we are not using in the code.

**Review by SaminYeasar — APPROVED** [review]

**Review by sordonia — APPROVED** [review]
